### ISSUE TITLE
Provide a mode where Stacked Borrows is enabled but borrowck isn't.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/miri#1751]
We could have tests that are 100% safe code, with two revisions:
* one of them is a compile-fail revision, with a borrowck error
* the other checks that miri with Stacked Borrows enabled but borrowck disabled, detects UB caused by the situation that borrowck disallows when enabled

Assuming miri can override queries, it should be easy to just make `borrowck` a noop (would be even cooler to turn errors into warnings, but idk if that's possible).

cc @oli-obk @RalfJung

**Comment by bjorn3** [comment]
The `mir_borrowck` function returns various things: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/query/struct.BorrowCheckResult.html

**Comment by RalfJung** [comment]
Yeah that sounds neat. :D Not sure how Miri would override queries though.

Also, of course there are programs that are fine by Stacked Borrows but still rejected by borrowck -- as usual in static analysis.

**Comment by oli-obk** [comment]
Oh heh. I'm sure this will cause ICEs, but that's still a fun experiment.

**Comment by saethlin** [comment]
I was reminded today in a conversation that Stacked Borrows base tags for stack memory are always `Unique`, regardless of whether the allocation is created with `let` or `let mut`, and that the only reason this program doesn't work is that it doesn't get past borrowck:
```rust
fn main() {
    unsafe {
        let thing = 1;
        let ptr = core::ptr::addr_of_mut!(thing);
        ptr.write(2);
    }
}
```
It seems likely that this would be well-defined according to Stacked Borrows.

**Comment by RalfJung** [comment]
Yeah, Stacked Borrows does not intend to give `let mut` any semantic meaning. Same for Tree Borrows. But if we want to actually have semantic constraints here, I think relying on borrowck is fine.
